enterprise/server/crypter_service/crypter_service.go

```diff
@@ -629,6 +641,164 @@ func (c *Crypter) NewDecryptor(ctx context.Context, digest *repb.Digest, r io.Re
 	return c.newDecryptorWithChunkSize(ctx, digest, r, em, u.GetGroupID(), plainTextChunkSize)
 }
 
+type encryptionKeyVersionWithGroupID struct {
+	GroupID string
+	tables.EncryptionKeyVersion
+}
+
+func (c *Crypter) reencryptKey(ctx context.Context, ekv *encryptionKeyVersionWithGroupID) error {
+	bbmk, err := c.kms.FetchMasterKey()
+	if err != nil {
+		return err
+	}
+
+	gmk, err := c.kms.FetchKey(ekv.GroupKeyURI)
+	if err != nil {
+		return err
+	}
+
+	masterKeyPortion, err := bbmk.Decrypt(ekv.MasterEncryptedKey, nil)
+	if err != nil {
+		return err
+	}
+	groupKeyPortion, err := gmk.Decrypt(ekv.GroupEncryptedKey, []byte(ekv.GroupID))
+	if err != nil {
+		return err
+	}
+	encMasterKeyPortion, err := bbmk.Encrypt(masterKeyPortion, nil)
+	if err != nil {
+		return err
+	}
+	encGroupKeyPortion, err := gmk.Encrypt(groupKeyPortion, []byte(ekv.GroupID))
+	if err != nil {
+		return err
+	}
+
+	q := `
+		UPDATE "EncryptionKeyVersions"
+		SET master_encrypted_key = ?,
+			group_encrypted_key = ?,
+			last_encryption_attempt_at_usec = ?,
+			last_encrypted_at_usec = ?
+		WHERE encryption_key_id = ? AND version = ?
+	`
+	now := c.clock.Now()
+	args := []interface{}{encMasterKeyPortion, encGroupKeyPortion, now.UnixMicro(), now.UnixMicro(), ekv.EncryptionKeyID, ekv.Version}
+	if err := c.dbh.DB(ctx).Exec(q, args...).Error; err != nil {
+		return err
+	}
+
+	log.Infof("Successfully re-encrypted key %q version %d", ekv.EncryptionKeyID, ekv.Version)
+
+	return nil
+}
+
+func (c *Crypter) keyRencryptorIteration(cutoff time.Time) error {
+	queryKeys := func() ([]*encryptionKeyVersionWithGroupID, error) {
+		ctx, cancel := context.WithTimeout(c.env.GetServerContext(), keyReencryptListQueryTimeout)
+		defer cancel()
+		q := `
```

Comment:
I'm not concerned with the # of keys since it should be proportional to the # of customers using encryption. That gives us a lot of room to scale. 

Rotation is configured in the external KMS system, so in order to figure out whether a key needs rotation would also require some kind of background process to reach out to the KMS for each key and check the rotation settings. I figured it would be simpler to just re-encrypt blindly.